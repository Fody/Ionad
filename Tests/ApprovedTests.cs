#if(DEBUG)

using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.Build.Utilities;
using NUnit.Framework;

[TestFixture]
[UseReporter(typeof(DiffReporter))]
public class ApprovedTests
{
    [Test]
    public void ILDasm()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath));
    }

    private static string Decompile(string assemblyPath)
    {
        var exePath = GetPathToILDasm();
        var process = Process.Start(new ProcessStartInfo(exePath, "\"" + assemblyPath + "\" /text /linenum")
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        });

        var projectFolder = Path.GetFullPath(Path.GetDirectoryName(assemblyPath) + "\\..\\..\\..").Replace("\\", "\\\\");
        projectFolder = Char.ToLower(projectFolder[0]) + projectFolder.Substring(1) + "\\\\";

        process.WaitForExit(10000);
        return string.Join(Environment.NewLine,
            Regex.Split(process.StandardOutput.ReadToEnd(), Environment.NewLine)
                .Where(l => !l.StartsWith("// Image base:"))
                .Select(l => l.Replace(projectFolder, ""))
        );
    }

    private static string GetPathToILDasm()
    {
        var path = Path.Combine(ToolLocationHelper.GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion.Version40), @"bin\NETFX 4.0 Tools\ildasm.exe");
        if (!File.Exists(path))
            path = path.Replace("v7.0", "v8.0");
        if (!File.Exists(path))
            Assert.Ignore("ILDasm could not be found");
        return path;
    }
}

#endif