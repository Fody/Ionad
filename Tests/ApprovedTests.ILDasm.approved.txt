
//  Microsoft (R) .NET Framework IL Disassembler.  Version 4.0.30319.1
//  Copyright (c) Microsoft Corporation.  All rights reserved.



// Metadata version: v4.0.30319
.assembly extern mscorlib
{
  .publickeytoken = (B7 7A 5C 56 19 34 E0 89 )                         // .z\V.4..
  .ver 4:0:0:0
}
.assembly extern retargetable mscorlib as mscorlib_1
{
  .publickeytoken = (7C EC 85 D7 BE A7 79 8E )                         // |.....y.
  .ver 2:0:5:0
}
.assembly AssemblyToProcess
{
  .custom instance void [mscorlib]System.Runtime.CompilerServices.CompilationRelaxationsAttribute::.ctor(int32) = ( 01 00 08 00 00 00 00 00 ) 
  .custom instance void [mscorlib]System.Runtime.CompilerServices.RuntimeCompatibilityAttribute::.ctor() = ( 01 00 01 00 54 02 16 57 72 61 70 4E 6F 6E 45 78   // ....T..WrapNonEx
                                                                                                             63 65 70 74 69 6F 6E 54 68 72 6F 77 73 01 )       // ceptionThrows.
  .custom instance void [mscorlib]System.Runtime.Versioning.TargetFrameworkAttribute::.ctor(string) = ( 01 00 1A 2E 4E 45 54 46 72 61 6D 65 77 6F 72 6B   // ....NETFramework
                                                                                                        2C 56 65 72 73 69 6F 6E 3D 76 34 2E 30 01 00 54   // ,Version=v4.0..T
                                                                                                        0E 14 46 72 61 6D 65 77 6F 72 6B 44 69 73 70 6C   // ..FrameworkDispl
                                                                                                        61 79 4E 61 6D 65 10 2E 4E 45 54 20 46 72 61 6D   // ayName..NET Fram
                                                                                                        65 77 6F 72 6B 20 34 )                            // ework 4

  // --- The following custom attribute is added automatically, do not uncomment -------
  //  .custom instance void [mscorlib]System.Diagnostics.DebuggableAttribute::.ctor(valuetype [mscorlib]System.Diagnostics.DebuggableAttribute/DebuggingModes) = ( 01 00 07 01 00 00 00 00 ) 

  .hash algorithm 0x00008004
  .ver 0:0:0:0
}
.module AssemblyToProcess.dll
// MVID: {4A26074C-C772-49C7-A53F-47DEBB7C9D41}
.imagebase 0x00400000
.file alignment 0x00000200
.stackreserve 0x00100000
.subsystem 0x0003       // WINDOWS_CUI
.corflags 0x00000001    //  ILONLY


// =============== CLASS MEMBERS DECLARATION ===================

.class public auto ansi beforefieldinit ClassWithGenericUsage
       extends [mscorlib]System.Object
{
  .method public hidebysig instance void 
          Method() cil managed
  {
    // Code size       8 (0x8)
    .maxstack  0
    .language '{3F5162F8-07C6-11D3-9053-00C04FA302A1}', '{994B45C4-E6E9-11D2-903F-00C04FA302A1}', '{5A869D0B-6611-11D3-BD2A-0000F80849BD}'
    .line 4,4 : 5,6 'AssemblyToProcess\\ClassWithGenericUsage.cs'
    IL_0000:  nop
    .line 5,5 : 9,44 ''
    IL_0001:  call       void class StaticGenericReplacement`1<string>::SomeMethod()
    IL_0006:  nop
    .line 6,6 : 5,6 ''
    IL_0007:  ret
  } // end of method ClassWithGenericUsage::Method

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method ClassWithGenericUsage::.ctor

} // end of class ClassWithGenericUsage

.class public auto ansi beforefieldinit StaticBasic
       extends [mscorlib]System.Object
{
  .method public hidebysig static void  SomeMethod() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    .line 6,6 : 5,6 'AssemblyToProcess\\StaticBasic.cs'
    IL_0000:  nop
    .line 7,7 : 9,45 ''
    IL_0001:  newobj     instance void [mscorlib]System.NotImplementedException::.ctor()
    IL_0006:  throw
  } // end of method StaticBasic::SomeMethod

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method StaticBasic::.ctor

} // end of class StaticBasic

.class public auto ansi beforefieldinit StaticBasicReplacementWithBrokenMethod
       extends [mscorlib]System.Object
{
  .method public hidebysig instance void 
          SomeMethod() cil managed
  {
    // Code size       2 (0x2)
    .maxstack  0
    .line 7,7 : 5,6 'AssemblyToProcess\\StaticBasicReplacementWithBrokenMethod.cs'
    IL_0000:  nop
    .line 8,8 : 5,6 ''
    IL_0001:  ret
  } // end of method StaticBasicReplacementWithBrokenMethod::SomeMethod

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method StaticBasicReplacementWithBrokenMethod::.ctor

} // end of class StaticBasicReplacementWithBrokenMethod

.class public auto ansi beforefieldinit ClassWithBrokenReplacement
       extends [mscorlib]System.Object
{
  .method public hidebysig instance void 
          Method() cil managed
  {
    // Code size       8 (0x8)
    .maxstack  0
    .line 4,4 : 5,6 'AssemblyToProcess\\ClassWithBrokenReplacement.cs'
    IL_0000:  nop
    .line 5,5 : 9,34 ''
    IL_0001:  call       void StaticBasic::SomeMethod()
    IL_0006:  nop
    .line 6,6 : 5,6 ''
    IL_0007:  ret
  } // end of method ClassWithBrokenReplacement::Method

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method ClassWithBrokenReplacement::.ctor

} // end of class ClassWithBrokenReplacement

.class public auto ansi beforefieldinit StaticGeneric`1<T>
       extends [mscorlib]System.Object
{
  .method public hidebysig static void  SomeMethod() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    .line 6,6 : 5,6 'AssemblyToProcess\\StaticGeneric.cs'
    IL_0000:  nop
    .line 7,7 : 9,45 ''
    IL_0001:  newobj     instance void [mscorlib]System.NotImplementedException::.ctor()
    IL_0006:  throw
  } // end of method StaticGeneric`1::SomeMethod

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method StaticGeneric`1::.ctor

} // end of class StaticGeneric`1

.class public auto ansi beforefieldinit StaticGenericReplacement`1<T>
       extends [mscorlib]System.Object
{
  .method public hidebysig static void  SomeMethod() cil managed
  {
    // Code size       2 (0x2)
    .maxstack  0
    .line 7,7 : 5,6 'AssemblyToProcess\\StaticGenericReplacement.cs'
    IL_0000:  nop
    .line 8,8 : 5,6 ''
    IL_0001:  ret
  } // end of method StaticGenericReplacement`1::SomeMethod

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method StaticGenericReplacement`1::.ctor

} // end of class StaticGenericReplacement`1

.class public abstract auto ansi sealed beforefieldinit StaticWithGenericMethodReplacement
       extends [mscorlib]System.Object
{
  .method public hidebysig static !!T  Method<T>() cil managed
  {
    // Code size       15 (0xf)
    .maxstack  1
    .locals init ([0] !!T CS$1$0000,
             [1] !!T CS$0$0001)
    .line 7,7 : 5,6 'AssemblyToProcess\\StaticWithGenericMethodReplacement.cs'
    IL_0000:  nop
    .line 8,8 : 9,27 ''
    IL_0001:  ldloca.s   CS$0$0001
    IL_0003:  initobj    !!T
    IL_0009:  ldloc.1
    IL_000a:  stloc.0
    IL_000b:  br.s       IL_000d

    .line 9,9 : 5,6 ''
    IL_000d:  ldloc.0
    IL_000e:  ret
  } // end of method StaticWithGenericMethodReplacement::Method

} // end of class StaticWithGenericMethodReplacement

.class public auto ansi beforefieldinit ClassWithDateTime
       extends [mscorlib]System.Object
{
  .method public hidebysig instance valuetype [mscorlib]System.DateTime 
          GetDateTime() cil managed
  {
    // Code size       11 (0xb)
    .maxstack  1
    .locals init ([0] valuetype [mscorlib]System.DateTime CS$1$0000)
    .line 6,6 : 5,6 'AssemblyToProcess\\ClassWithDateTime.cs'
    IL_0000:  nop
    .line 7,7 : 9,29 ''
    IL_0001:  call       valuetype [mscorlib]System.DateTime DateTimeReplacement::get_Now()
    IL_0006:  stloc.0
    IL_0007:  br.s       IL_0009

    .line 8,8 : 5,6 ''
    IL_0009:  ldloc.0
    IL_000a:  ret
  } // end of method ClassWithDateTime::GetDateTime

  .method public hidebysig specialname instance valuetype [mscorlib]System.DateTime 
          get_SomeProperty() cil managed
  {
    // Code size       11 (0xb)
    .maxstack  1
    .locals init ([0] valuetype [mscorlib]System.DateTime CS$1$0000)
    .line 10,10 : 40,41 ''
    IL_0000:  nop
    .line 10,10 : 42,62 ''
    IL_0001:  call       valuetype [mscorlib]System.DateTime DateTimeReplacement::get_Now()
    IL_0006:  stloc.0
    IL_0007:  br.s       IL_0009

    .line 10,10 : 63,64 ''
    IL_0009:  ldloc.0
    IL_000a:  ret
  } // end of method ClassWithDateTime::get_SomeProperty

  .method public hidebysig instance valuetype [mscorlib]System.DateTime 
          MissingReplacement() cil managed
  {
    // Code size       11 (0xb)
    .maxstack  1
    .locals init ([0] valuetype [mscorlib]System.DateTime CS$1$0000)
    .line 13,13 : 5,6 ''
    IL_0000:  nop
    .line 14,14 : 9,31 ''
    IL_0001:  call       valuetype [mscorlib]System.DateTime [mscorlib]System.DateTime::get_Today()
    IL_0006:  stloc.0
    IL_0007:  br.s       IL_0009

    .line 15,15 : 5,6 ''
    IL_0009:  ldloc.0
    IL_000a:  ret
  } // end of method ClassWithDateTime::MissingReplacement

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method ClassWithDateTime::.ctor

  .property instance valuetype [mscorlib]System.DateTime
          SomeProperty()
  {
    .get instance valuetype [mscorlib]System.DateTime ClassWithDateTime::get_SomeProperty()
  } // end of property ClassWithDateTime::SomeProperty
} // end of class ClassWithDateTime

.class public auto ansi beforefieldinit ClassWithGenericMethodUsage
       extends [mscorlib]System.Object
{
  .method public hidebysig instance void 
          Method() cil managed
  {
    // Code size       8 (0x8)
    .maxstack  1
    .line 4,4 : 5,6 'AssemblyToProcess\\ClassWithGenericMethodUsage.cs'
    IL_0000:  nop
    .line 5,5 : 9,50 ''
    IL_0001:  call       !!0 StaticWithGenericMethodReplacement::Method<string>()
    IL_0006:  pop
    .line 6,6 : 5,6 ''
    IL_0007:  ret
  } // end of method ClassWithGenericMethodUsage::Method

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method ClassWithGenericMethodUsage::.ctor

} // end of class ClassWithGenericMethodUsage

.class public abstract auto ansi sealed beforefieldinit DateTimeReplacement
       extends [mscorlib]System.Object
{
  .method public hidebysig specialname static 
          valuetype [mscorlib]System.DateTime 
          get_Now() cil managed
  {
    // Code size       19 (0x13)
    .maxstack  3
    .locals init ([0] valuetype [mscorlib]System.DateTime CS$1$0000)
    .line 7,7 : 38,39 'AssemblyToProcess\\DateTimeReplacement.cs'
    IL_0000:  nop
    .line 7,7 : 40,73 ''
    IL_0001:  ldc.i4     0x7ba
    IL_0006:  ldc.i4.1
    IL_0007:  ldc.i4.s   13
    IL_0009:  newobj     instance void [mscorlib]System.DateTime::.ctor(int32,
                                                                        int32,
                                                                        int32)
    IL_000e:  stloc.0
    IL_000f:  br.s       IL_0011

    .line 7,7 : 74,75 ''
    IL_0011:  ldloc.0
    IL_0012:  ret
  } // end of method DateTimeReplacement::get_Now

  .property valuetype [mscorlib]System.DateTime
          Now()
  {
    .get valuetype [mscorlib]System.DateTime DateTimeReplacement::get_Now()
  } // end of property DateTimeReplacement::Now
} // end of class DateTimeReplacement

.class public auto ansi beforefieldinit StaticWithGenericMethod
       extends [mscorlib]System.Object
{
  .method public hidebysig static !!T  Method<T>() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    .line 6,6 : 5,6 'AssemblyToProcess\\StaticWithGenericMethod.cs'
    IL_0000:  nop
    .line 7,7 : 9,45 ''
    IL_0001:  newobj     instance void [mscorlib]System.NotImplementedException::.ctor()
    IL_0006:  throw
  } // end of method StaticWithGenericMethod::Method

  .method public hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  1
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method StaticWithGenericMethod::.ctor

} // end of class StaticWithGenericMethod


// =============================================================

// *********** DISASSEMBLY COMPLETE ***********************
